### ISSUE TITLE
Chain storage separation

### ISSUE THREAD

**Issue by cpfiffer** [TuringLang/MCMCChains.jl#30]
Currently a `Chains` object stores a bunch of data in a more or less equal format. For example, `describe(chain)` might return something like this from a model generated with Turing:

```
Iterations = 1:1000
Thinning interval = 1
Chains = 1
Samples per chain = 1000

Empirical Posterior Estimates:
              Mean          SD       Naive SE      MCSE        ESS
       m  1.1715753065  0.98994393 0.031304776 0.082612482  143.591870
  lf_num  0.0060000000  0.18973666 0.006000000 0.006000000 1000.000000
       s  2.3994084547  2.79358857 0.088341027 0.233425460  143.228171
 elapsed  0.0027014705  0.03728031 0.001178907 0.001703629  478.860457
 epsilon  0.7113239044  0.56295227 0.017802114 0.041734566  181.949702
      lp -5.9499451689  1.50461250 0.047580025 0.138894319  117.349264
eval_num 12.6320000000 11.46507846 0.362557615 0.600414301  364.629671
  lf_eps  0.7113239044  0.56295227 0.017802114 0.041734566  181.949702

Quantiles:
               2.5%          25.0%        50.0%        75.0%        97.5%
       m  -0.79725744431  0.657272530  1.148128006  1.647442905  3.351685253
  lf_num   0.00000000000  0.000000000  0.000000000  0.000000000  0.000000000
       s   0.57576571615  1.096100078  1.616797620  2.460184551 10.796821566
 elapsed   0.00027849772  0.000659921  0.000758211  0.000969643  0.003740704
 epsilon   0.17805198440  0.601473075  0.601473075  0.659085622  2.040240522
      lp -10.48666390455 -6.448987120 -5.382348703 -4.953985494 -4.634529304
eval_num   4.00000000000 10.000000000 10.000000000 10.000000000 46.000000000
  lf_eps   0.17805198440  0.601473075  0.601473075  0.659085622  2.040240522
```

Some of those values are parameters, and some are other variables generated during sampling. I would like to see a scenario where `m` and `s` are visually separated (and sorted alphabetically, see #24) from values like `lp`, `elapsed`, and `epsilon`. 

@yebai suggested using an underscore prefix to denote "internal" values like `lp`, such that those values with underscores at the start of their names can be placed separately from the parameters sampled. 

I'm not sure I like this idea -- I would prefer a general dictionary format (though maybe that's too slow, thoughts?) where you can specify "sections" of a `ChainSummary` object to report. This would work for cases where the `ChainSummary` object is used for other things like printing out diagnostics (`heideldiag`) or what have you.

As an example, imagine storing all the parameters in `chain.variable_dict[:Parameters]` and all the non-parameters in `chain.variable_dict[:Internals]`. Then, when printing a `ChainSummary`, we might see something more like:

```
Quantiles:
               2.5%          25.0%        50.0%        75.0%        97.5%
Parameters:
       m  -0.79725744431  0.657272530  1.148128006  1.647442905  3.351685253
       s   0.57576571615  1.096100078  1.616797620  2.460184551 10.796821566

Internals:
  lf_num   0.00000000000  0.000000000  0.000000000  0.000000000  0.000000000
 elapsed   0.00027849772  0.000659921  0.000758211  0.000969643  0.003740704
 epsilon   0.17805198440  0.601473075  0.601473075  0.659085622  2.040240522
      lp -10.48666390455 -6.448987120 -5.382348703 -4.953985494 -4.634529304
eval_num   4.00000000000 10.000000000 10.000000000 10.000000000 46.000000000
  lf_eps   0.17805198440  0.601473075  0.601473075  0.659085622  2.040240522
```

I wanted to get thoughts from not only the Turing people but also anyone else who uses MCMCChains, like @goedman. It's a very rudimentary idea, but I'm open to suggestions on how to allow for the separate printing and possibly processing. An adjustment like this might make it easier for Turing to grab parameter estimates after `NUTS`, so that adaptation period samples aren't included in the chain.

Comments encouraged.

**Comment by goedman** [comment]
One other point I have noticed is that if you call describe on a single variable you get a single column not showing the sd, e.g. describe(chain[:m]). If this is not clear I'll create an example.

Subsetting variables in Turing is in general a bit harder than in Mamba and Stan, e.g. chain.value[:,[:m, :s],:] (vs. chain.value[:,2,:] currently).

**Comment by goedman** [comment]
@cpfiffer Forgot to mention an important aspect, I think your proposed layout looks great! I think we should include the posterior summary for the parameters and limit the number of digits. Maybe make the internal part optional?

**Comment by cpfiffer** [comment]
I think it's a good idea to limit the digits, and to include some kind of `describe(chain; internals = false)` functionality. I'm looking into how to index better. Turing has its own version of an `AbstractChains` object, and that indexes a lot more intuitively, i.e., `chain[:m]`, while the default MCMCChain struct uses the very wordy `chain[:,1,:]` which I am not too fond of.

**Comment by trappmartin** [comment]
I started changing the internal data structure to use an AxisArray. Doing so would solve several other issue I'm having with the current design. It seems to me that a combination of your proposal with AxisArrays would be the most general solution.

Consider the following example:

```julia
# data storage
vals = Dict{Symbol, AxisArray}()

# shared parameters
chains = Axis{:chain}(1:3)
ranges = Axis{:range}(1:100:10000)

# parameters
pnames = Axis{:name}(["m", "s"])
parameters = cat(randn(length(chains), length(1:100:10000)) .+ 1, randn(length(chains), length(1:100:10000)), dims=3)

# internal parameters
inames = Axis{:name}(["epsilon"])
internals = rand(length(chains), length(1:100:10000), 1)

# scores
snames = Axis{:name}(["lp"])
scores = log.(rand(length(chains), length(1:100:10000), 1))

vals[:parameters] = AxisArray(parameters, chains, ranges, pnames)
vals[:internalparameters] = AxisArray(internals, chains, ranges, inames)
vals[:scores] = AxisArray(scores, chains, ranges, snames)
```

This way we would not need the additional transformation functions, `getindex` and `setindex!` functions anymore. The Chains type would reduce to something like this:

```julia
struct Chains{T<:Real} <: AbstractChains
    logevidence::T
    value::Dict{Symbol, AxisArray{Union{Missing, T}, 3}}
    uniquenames::Dict{Symbol, String}
end
```

I think this would be a good idea.

**Comment by trappmartin** [comment]
#23

**Comment by xukai92** [comment]
For my work on improving adaptation (https://github.com/TuringLang/Turing.jl/pull/641), we actually need to monitor internal parameters and model parameters separately. I'm doing this (in my local branch) by adding `_` to internal parameters. Is this OK on this side?

**Comment by cpfiffer** [comment]
Yes, that won't mess anything up. The way I've decided to implement this is to give `Chains` objects "sections". Sections map a list of variable names to a category of data. For example, `:parameter` might map to `[:alpha, :beta]`, while `:internals` might map to `[:lp, :nsamples]`, etc. Most commonly, I would expect `Chains` to have two sections (`:paramaters` and `:internals`). 

Sadly the API for this isn't super stable at the moment, but not much will change after the changeover. I'll go in and later and update everyone's code later to make sure that we're using the mapping right on the Turing side.

**Comment by goedman** [comment]
Hi Cameron, looking through other parts of MCMCChain (currently fileio.jl and next modelstats.jl), do you know what the status is of those? Or the plan? 

Fileio.jl came directly from Mamba.jl and would work if the stdlib Serialization read and write would be imported. In StanMCMCChain I use serialize/deserialize directly but it would be nicer if this would work in MCMCChain. I can make a PR from my MCMCChain fork.

Modelchains.jl currently is not included (commented out) in MCMCChain.jl. I'm quite interested in it, also for StatisticalRethinking. Ok if I take a look ?

**Comment by cpfiffer** [comment]
I've added a pull request to get read/write back up and running, see #32. 

As to `modelchains.jl` and `modelstats.jl`, I hadn't even begun to look at those. It looks like they mostly support helpful functions for post-sampling exploration, like `predict` and `logpdf`. I'd like to get those working with straight `Chains` objects, as I don't really see the need to have two different subtypes of `AbstractChains` objects if we can just get whatever functions people want working on `Chains` objects.

It's fine if you take a look! There's a lot of stuff moving around as far as the internals are concerned while I'm switching everything over to `AxisArrays` (#20) so just be aware that the API might slip away from you. 

In that case, I'd be happy to update whatever you'd worked on up to that point. Alternatively, if you'd prefer to open an issue with a functionality wish list I can stick that on my list.

**Comment by goedman** [comment]
Thanks, I’ll take a look and get back to you.

**Comment by cpfiffer** [comment]
Closed by #33, for now.
